File: src/tw/rmstudio/mt/java/course/mcu/ds/ex1/sorting/MergeSort.java
package tw.rmstudio.mt.java.course.mcu.ds.ex1.sorting;

import java.util.Arrays;

public class MergeSort extends SortingAlgorithm {
    private int lastConditionCount = -1, lastSwapCount = -1;

    @Override
    int[] sort(int[] randomNumbers) {
        int conditionCounter = 0, swapCounter = 0;

        int length = randomNumbers.length;
        int[] sortedNumbers = new int[length], sourceNumbers = Arrays.copyOf(randomNumbers, length);
        int block, index;

        for (block = 1; block < length * 2 /* ?! */ ; block *= 2) {
            for (index = 0; index < length; index += block * 2) {
                int center = Math.min(index + block, length), upper = Math.min(index + block * 2, length);
                int k = index, indexA = index, indexB = center;

                while (indexA < center && indexB < upper) {
                    sortedNumbers[k++] = sourceNumbers[indexA] < sourceNumbers[indexB] ? sourceNumbers[indexA++] : sourceNumbers[indexB++];
                    ++conditionCounter;
                }

                while (indexA < center) sortedNumbers[k++] = sourceNumbers[indexA++];
                while (indexB < upper) sortedNumbers[k++] = sourceNumbers[indexB++];
            }

            ++swapCounter;
            int[] swapTempArray = sourceNumbers;
            sourceNumbers = sortedNumbers;
            sortedNumbers = swapTempArray;
        }
        sortedNumbers = sourceNumbers;

        lastConditionCount = conditionCounter;
        lastSwapCount = swapCounter;
        return sortedNumbers;
    }

    @Override
    public String getAlgorithmName() {
        return "Merge Sort";
    }

    @Override
    int getConditionCount() {
        return lastConditionCount;
    }

    @Override
    public int getSwapCount() {
        return lastSwapCount;
    }
}

File: src/tw/rmstudio/mt/java/course/mcu/ds/ex1/sorting/MergeSort.kt
package tw.rmstudio.mt.java.course.mcu.ds.ex1.sorting

import java.util.*

class MergeSort : SortingAlgorithm() {
    override var conditionCount = -1
        private set
    override var swapCount = -1
        private set

    override val algorithmName: String
        get() = "Merge Sort"

    override fun sort(randomNumbers: IntArray): IntArray {
        var conditionCounter = 0
        var swapCounter = 0

        val length = randomNumbers.size
        var sortedNumbers = IntArray(length)
        var sourceNumbers = Arrays.copyOf(randomNumbers, length)
        var block = 1
        var index: Int

        while (block < length * 2) {
            index = 0
            while (index < length) {
                val center = Math.min(index + block, length)
                val upper = Math.min(index + block * 2, length)
                var k = index
                var indexA = index
                var indexB = center

                while (indexA < center && indexB < upper) {
                    sortedNumbers[k++] = if (sourceNumbers[indexA] < sourceNumbers[indexB]) sourceNumbers[indexA++] else sourceNumbers[indexB++]
                    ++conditionCounter
                }

                while (indexA < center) sortedNumbers[k++] = sourceNumbers[indexA++]
                while (indexB < upper) sortedNumbers[k++] = sourceNumbers[indexB++]
                index += block * 2
            }

            ++swapCounter
            val swapTempArray = sourceNumbers
            sourceNumbers = sortedNumbers
            sortedNumbers = swapTempArray
            block *= 2
        }
        sortedNumbers = sourceNumbers

        conditionCount = conditionCounter
        swapCount = swapCounter
        return sortedNumbers
    }
}
